### DISCUSSION TITLE
Headless 컴포넌트의 하위 컴포넌트 타입 지정 방식

### DISCUSSION THREAD

**Discussion by hemudi** [Co-Studo/Co-Studo-front#25] [General]
---

[PR : [Hemdi] Radio Component 생성](https://github.com/Co-Studo/Co-Studo-front/pull/24)
원래 PR에 작성했다가 PR 내용이랑 의논 내용은 가능하면 분리해서 기록하는게 좋을 것 같아서 discussion 을 팠습니다!

---

[[ESLint / TypeScript] Optionl 속성의 default value 문제 + React.FC 사용 논의](https://github.com/Co-Studo/Co-Studo-front/discussions/22)

위의 Discussion 을 통해 최종적으로 컴포넌트들의 타입을 React.FC으로 지정하기로 결정한 뒤 적용을 해보니 Headless 컴포넌트에 하위 컴포넌트를 추가하는 부분에서 type 에러가 발생했습니다.

![image](https://user-images.githubusercontent.com/34249911/192461689-3d0cfe16-038e-4a02-ab8b-3ffffb237fcc.png)
![error](https://user-images.githubusercontent.com/34249911/192461606-7b43d34a-a579-4be6-80a4-a4c7c1b072e3.PNG)

```tsx
type RadioGroupChildren = {
  Title: React.FC<TitleProps>;
  Option: React.FC<OptionProps>;
};

const RadioGroup: React.FC<RadioGroupProps> & RadioGroupChildren = ({}) => {}
```

그래서 위와 같이 하위 컴포넌트의 타입을 따로 정의해서 추가해줘서 [오류는 해결](https://github.com/Co-Studo/Co-Studo-front/commit/11566fae5e0db91300f279fd2b0f2e670e94707a#diff-21e61eb3de84152155a0f307cfe48463ad89adf2b1a647b7cfb1d59ad469819eR33)은 했는데 **적절한 네이밍**인지,
이게 최선의 해결 방법인지 의문이 들었습니다.
그리고 모든 Headless 컴포넌트에서 방식이나 타입 네이밍을 통일 하는게 좋을 것 같아서 함께 의논을 해보고 싶습니다!

제가 위 방식이 최선의 해결 방법인지 의문이 든 포인트는 하위 컴포넌트는 추가되거나 변경, 삭제 될 수 있는 요소인데 그럴 때마다 계속 type을 수정해줘야 한다는 점 + `React.FC<RadioGroupProps> & RadioGroupChildren` 이 부분의 가독성이 그렇게 좋아보이지 않다는 점입니다.

그래서 다른 방법이 없을까 싶어서 찾아보니 [stackoverflow ](https://stackoverflow.com/a/60021394)에 아래와 같은 방법도 있더라구요.

```tsx
const RadioGroupComponent: React.FC<RadioGroupProps> = ({}) => {}
const Option: React.FC<RadioOptionProps> = ({}) => {}
const Title: React.FC<TitleProps> = ({ children, ...props }) => ({}) => {}

const RadioGroup = Object.assign(RadioGroupComponent, { Option, Title });
export default RadioGroup;
```

type을 지정해주지 않고 Object.assign을 이용해 객체를 합쳐 export 하는 방식입니다. 
괜찮아 보이긴 한데 만약 이 방식을 적용해야 한다면 현재의 모든 Headless 컴포넌트에서 방식을 다 통일하는 작업을 해야해서
그냥 type 변동으로 인한 코드 수정을 감수하고 현재의 방식으로 하는 게 나은 것 같기도 합니다. @ _@)

**Comment by healtheloper** [comment]
햄디가 올려주신 Discussion 의 [추가 댓글로 단 부분](https://github.com/Co-Studo/Co-Studo-front/discussions/22#discussioncomment-3739881)에 **Compound 에서 타입이 애매해진다** 라는 부분에 동의만 하고 넘어간 부분이 바로 문제로 발생했군여 

처음 해당 문제를 보았을 때에는 `RadioGroupProps` 에 `RadioGroupChildren` 을 묶으신 부분을 풀어서 넣는 방법을 생각했었어요
그러면 `React.FC<RadioGroupProps> & RadioGroupChildren` -> `React.FC<RadioGroupProps>` 로 표현할 수 있지 않을까요?

```tsx
type RadioGroupProps = {
  ...
  Title: React.FC<TitleProps>
  ...
}
```

assign 하는 방식도 좋아보여요 만약 compound 방식의 컴포넌트 설계 시 해당  방법으로 합치자는 컨벤션만 맞추면 된다고 생각해요

> 하위 컴포넌트는 추가되거나 변경, 삭제 될 수 있는 요소인데 그럴 때마다 계속 type을 수정해줘야 한다는 점

해당 부분은 Title, Option 같은 컴포넌트를 얘기하는 걸까요? 적절한 예시가 떠오르지않네용
TitleProps 로 넣어주면 TitleProps 가 변경되어도 대응되지않을까 란 생각만 들었어요

**Reply by hemudi** [reply]
![image](https://user-images.githubusercontent.com/34249911/192482655-5237f0d3-9fb9-4f37-84aa-1d3d7172bcc2.png)
그렇지 않아도 저도 처음에는 말씀해주신 것처럼 풀어서 넣는 방식을 제일 먼저 시도해봤었는데 그래도 여전히 같은 에러가 뜨더라구요. 혹시 제 환경에서만 그런걸까요?^ -ㅠ)

---

> 하위 컴포넌트는 추가되거나 변경, 삭제 될 수 있는 요소인데 그럴 때마다 계속 type을 수정해줘야 한다는 점

이 부분은 Title, Option 같은 컴포넌트 (또는 속성)을 말씀드린게 맞습니당!
그런데 Title의 내부의 변경이라기보다 컴포넌트의 존재 자체의 변경(?)에 대해서 고민을 했던거 같아요!

```tsx
type RadioGroupChildren = {
  Title: React.FC<TitleProps>;
  Option: React.FC<OptionProps>;
  Label: React.FC<LabelProps>;  // 2. 타입도 한번 추가해주고
};

const RadioGroup: React.FC<RadioGroupProps> & RadioGroupChildren = ({}) => {}
const Title: React.FC<TitleProps> = ({ children, ...props }) => ({}) => {}
const Option: React.FC<OptionProps> = ({}) => {}
const Label: React.FC<LabelProps> = ({}) => {}   // 1. Label 컴포넌트가 하나 새로 추가 된다면.

RadioGroup.Title = Title;
RadioGroup.Opton = Option;
RadioGroup.Label = Label;  // 3. 값도 넣어주고

export default RadioGroup;
```
예를 들어 위 처럼 Compound 시킬 컴포넌트 Label 이라는게 새로 추가 된다고 하면
RadioGroupChildren 타입에 해당 컴포넌트에 대한 타입을 추가 해줘야 하는데
타입을 찾고 추가해주고 하는 과정이 좀 번거로울 것 같더라구요!  
(너무 사소한 귀찮음일까요ㅋㅋㅋㅋㅋㅋㅋㅋ^///^)>)

여튼 그래서 원래는 그냥 새 컴포넌트를 추가하면 3 번의 작업만 신경써주면 됐는데 과정이 하나 추가 된 게 좀 번거로운 것 같았고!
저렇게 Compound 될 컴포넌트들의 type을 따로 신경써서 모아놓을 필요가 있나? 어차피 결국 다 같은 React.FC 타입인데 싶어서 좀 더 나은 해결법이 없나 고민이 되었던 것 같아요!

**Reply by healtheloper** [reply]
아, 생각해보니 React.FC 에 들어가는 Props 가 아니라서 오류가 나는게 맞겠네요 
props.children 은 있지만 props.Title 이 없으니 당연한거였..

예시를 보니 확실히 귀찮음이 느껴지는데 assign 하는 방식에 저는 한표.. 합니다
